sqlx-postgres/src/bind_iter.rs

```diff
@@ -0,0 +1,152 @@
+/// sqlx::query("insert into person(id, name, birthdate) select * from unnest($1, $2, $3)")
+///     .bind(people.iter().map(|p| p.id).bind_iter())
+///     .bind(people.iter().map(|p| &p.name).bind_iter())
+///     .bind(people.iter().map(|p| &p.birthdate).bind_iter())
+///     .execute(&mut conn)
+///     .await?;
+///
+/// # Ok(())
+/// # }
+/// ```
+pub trait PgBindIterExt: Iterator + Sized {
+    fn bind_iter(self) -> PgBindIter<Self>;
+}
+
+impl<I: Iterator + Sized> PgBindIterExt for I {
+    fn bind_iter(self) -> PgBindIter<I> {
+        PgBindIter(self)
+    }
+}
+
+impl<I> Type<Postgres> for PgBindIter<I>
+where
+    I: Iterator,
+    <I as Iterator>::Item: Type<Postgres> + PgHasArrayType,
+{
+    fn type_info() -> <Postgres as Database>::TypeInfo {
+        <I as Iterator>::Item::array_type_info()
+    }
+    fn compatible(ty: &PgTypeInfo) -> bool {
+        <I as Iterator>::Item::array_compatible(ty)
+    }
+}
+
+impl<'q, I> PgBindIter<I>
+where
+    I: Iterator,
+    <I as Iterator>::Item: Type<Postgres> + Encode<'q, Postgres>,
+{
+    fn encode_inner(
+        // need ownership to iterate
+        mut iter: I,
+        buf: &mut PgArgumentBuffer,
+    ) -> Result<IsNull, BoxDynError> {
+        let lower_size_hint = iter.size_hint().0;
+        let first = iter.next();
+        let type_info = first
+            .as_ref()
+            .and_then(Encode::produces)
+            .unwrap_or_else(<I as Iterator>::Item::type_info);
+
+        buf.extend(&1_i32.to_be_bytes()); // number of dimensions
+        buf.extend(&0_i32.to_be_bytes()); // flags
+
+        match type_info.0 {
+            PgType::DeclareWithName(name) => buf.patch_type_by_name(&name),
+            PgType::DeclareArrayOf(array) => buf.patch_array_type(array),
+
+            ty => {
+                buf.extend(&ty.oid().0.to_be_bytes());
+            }
+        }
+
+        let len_start = buf.len();
+        buf.extend(0_i32.to_be_bytes()); // len (unknown so far)
+        buf.extend(1_i32.to_be_bytes()); // lower bound
+
+        match first {
+            Some(first) => buf.encode(first)?,
+            None => return Ok(IsNull::No),
+        }
+
+        let mut count = 1_i32;
+        const MAX: usize = i32::MAX as usize - 1;
+
+        for value in (&mut iter).take(MAX) {
+            buf.encode(value)?;
+            count += 1;
+        }
+
+        const OVERFLOW: usize = i32::MAX as usize + 1;
+        if iter.next().is_some() {
+            let iter_size = std::cmp::max(lower_size_hint, OVERFLOW);
+            return Err(format!("encoded iterator is too large for Postgres: {iter_size}").into());
+        }
+
+        // set the length now that we know what it is.
+        buf[len_start..(len_start + 4)].copy_from_slice(&count.to_be_bytes());
+
+        Ok(IsNull::No)
+    }
+}
+
+impl<'q, I> Encode<'q, Postgres> for PgBindIter<I>
+where
+    // Clone is required for the encode_by_ref call since we can't iterate with a shared reference
+    I: Iterator + Clone,
```

Comment:
The plan I had for this was to use `Cell<Option<I>>` since the `Clone` bound could theoretically be hard to satisfy for some iterators.

Using `Cell` would make it `!Sync` and only able to used once, but those seem like pretty reasonable tradeoffs. I don't expect anyone to need to share this across threads or or need to bind the same iterator more than once (they could just use the same placeholder number to refer to the existing binding or create a new iterator).

It seems like maybe we should split `Encode` into two traits: one that encodes by-value and one by-reference, then a blanket impl of the former over the latter. The by-reference trait would be the one that most types would implement, but the by-value trait would be the one that all the APIs actually accept.

Reply:
Ok, I can make the change. I don't think it will affect the send-ness of the execution future so that should be fine.

I agree with splitting up the Encode trait. It would make this simpler.

Reply:
@abonander I've made the change to `Cell<Option<I>>`